pkg/backend/httpstate/token_source.go

```diff
@@ -130,11 +130,21 @@ func (ts *tokenSource) handleRequests(
 	}
 }
 
-func (ts *tokenSource) GetToken() (string, error) {
+func (ts *tokenSource) GetToken(ctx context.Context) (string, error) {
 	ch := make(chan tokenResponse)
-	ts.requests <- ch
-	resp := <-ch
-	return resp.token, resp.err
+
+	select {
+	case ts.requests <- ch:
+	case <-ctx.Done():
+		return "", ctx.Err()
+	}
+
+	select {
+	case resp := <-ch:
+		return resp.token, resp.err
+	case <-ctx.Done():
+		return "", ctx.Err()
+	}
```

Comment:
> optional: It's impossible for ch to produce a value before ts.requests <- ch succeeds, so it's possible to combine these two select blocks into one.

Unfortunately it isn't--the blocks need to be separate because the receive must happen after the send. If the blocks are combined, then we won't try to do the send without another block (or a loop).

Reply:
Derp. 🤦  You're right. My brain assumed a surrounding for loop.